Source/WTF/wtf/glib/RunLoopGLib.cpp

```diff
@@ -35,21 +36,68 @@
 #include <wtf/glib/ActivityObserver.h>
 #include <wtf/glib/RunLoopSourcePriority.h>
 
+#if HAVE(TIMERFD)
+#include <sys/timerfd.h>
+#include <time.h>
+#include <wtf/SystemTracing.h>
+#endif
+
 namespace WTF {
 
 typedef struct {
     GSource source;
     RunLoop* runLoop;
+#if HAVE(TIMERFD)
+    int timerFd;
+    struct itimerspec timerFdSpec;
+#endif
 } RunLoopSource;
 
 GSourceFuncs RunLoop::s_runLoopSourceFunctions = {
-    nullptr, // prepare
+    // prepare
+#if HAVE(TIMERFD)
+    [](GSource* source, int* timeout) -> gboolean {
+        auto& runLoopSource = *reinterpret_cast<RunLoopSource*>(source);
+
+        *timeout = -1;
+
+        if (runLoopSource.timerFd > -1) {
+            struct itimerspec timerFdSpec = { { 0, 0 }, { 0, 0 } };
+            int64_t readyTime = g_source_get_ready_time(source);
+
+            if (readyTime > -1) {
+                timerFdSpec.it_value.tv_sec = readyTime / G_USEC_PER_SEC;
+                timerFdSpec.it_value.tv_nsec = (readyTime % G_USEC_PER_SEC) * 1000L;
+            }
+
+            if (timerFdSpec.it_interval.tv_sec != runLoopSource.timerFdSpec.it_interval.tv_sec
+                || timerFdSpec.it_interval.tv_nsec != runLoopSource.timerFdSpec.it_interval.tv_nsec
+                || timerFdSpec.it_value.tv_sec != runLoopSource.timerFdSpec.it_value.tv_sec
+                || timerFdSpec.it_value.tv_nsec != runLoopSource.timerFdSpec.it_value.tv_nsec) {
+                runLoopSource.timerFdSpec = timerFdSpec;
+                timerfd_settime(runLoopSource.timerFd, TFD_TIMER_ABSTIME, &runLoopSource.timerFdSpec, nullptr);
+            }
+        }
+        return FALSE;
+    },
+#else
+    nullptr,
+#endif
     nullptr, // check
     // dispatch
     [](GSource* source, GSourceFunc callback, gpointer userData) -> gboolean
     {
-        if (g_source_get_ready_time(source) == -1)
+        gint64 readyTime = g_source_get_ready_time(source);
+        if (readyTime == -1)
             return G_SOURCE_CONTINUE;
+
+#if HAVE(TIMERFD) && USE(SYSPROF_CAPTURE)
+        if (readyTime > 0) {
+            gint64 lateness = g_get_monotonic_time() - readyTime;
+            WTFEmitSignpost(source, RunLoopSourceDispatch, "[%s] lateness=%ldµs", g_source_get_name(source), lateness);
```

Comment:
I'm worried that we're swamped with marks in sysprof. How about adding an env var to toggle this at runtime, defaulting to false. This we we can utilize the mechanism if needed, and avoid adding many new marks, if not necessary.